src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs

```diff
@@ -966,29 +1001,44 @@ private bool SatisfiesGetEnumeratorPattern(ref ForEachEnumeratorInfo.Builder bui
         /// <param name="warningsOnly">True if failures should result in warnings; false if they should result in errors.</param>
         /// <param name="diagnostics">Populated with binding diagnostics.</param>
         /// <returns>The desired method or null.</returns>
-        private MethodSymbol FindForEachPatternMethod(TypeSymbol patternType, string methodName, LookupResult lookupResult, bool warningsOnly, DiagnosticBag diagnostics, bool isAsync)
+        private MethodSymbol FindForEachPatternMethod(TypeSymbol patternType, string methodName, LookupResult lookupResult, bool warningsOnly, bool reportNoSuchMember, DiagnosticBag diagnostics, bool isAsync, bool viaExtensionMethod, out bool reportedError)
         {
+            reportedError = false;
             Debug.Assert(lookupResult.IsClear);
 
             // Not using LookupOptions.MustBeInvocableMember because we don't want the corresponding lookup error.
             // We filter out non-methods below.
             HashSet<DiagnosticInfo> useSiteDiagnostics = null;
-            this.LookupMembersInType(
-                lookupResult,
-                patternType,
-                methodName,
-                arity: 0,
-                basesBeingResolved: null,
-                options: LookupOptions.Default,
-                originalBinder: this,
-                diagnose: false,
-                useSiteDiagnostics: ref useSiteDiagnostics);
+
+            if (viaExtensionMethod)
+            {
+                this.LookupExtensionMethods(
+                    lookupResult,
+                    methodName,
+                    arity: 0,
+                    LookupOptions.IncludeExtensionMethods | LookupOptions.AllMethodsOnArityZero,
+                    useSiteDiagnostics: ref useSiteDiagnostics);
+            }
+            else
+            {
+                this.LookupMembersInType(
+                    lookupResult,
+                    patternType,
+                    methodName,
+                    arity: 0,
+                    basesBeingResolved: null,
+                    options: LookupOptions.Default,
+                    originalBinder: this,
+                    diagnose: false,
+                    useSiteDiagnostics: ref useSiteDiagnostics);
+            }
 
             diagnostics.Add(_syntax.Expression, useSiteDiagnostics);
 
             if (!lookupResult.IsMultiViable)
             {
-                ReportPatternMemberLookupDiagnostics(lookupResult, patternType, methodName, warningsOnly, diagnostics);
+                ReportPatternMemberLookupDiagnostics(lookupResult, patternType, methodName, warningsOnly, reportNoSuchMember, diagnostics);
+                // We don't set reportedError to true since we can get a more specific error message further up
```

Comment:
Please be more specific here.

Reply:
And this pattern seems wrong in general. Either we should be reporting lookup diagnostics, and suppressing future diagnostics, or we should not. I think this needs a refactor.

---
In reply to: [412344868](https://github.com/dotnet/roslyn/pull/43050#discussion_r412344868) [](ancestors = 412344868)

Reply:
Consider the following example:

```csharp
using System;
public class C
{
    public static void Main()
    {
        foreach (var i in new C())
        {
            Console.Write(i);
        }
    }
    public sealed class Enumerator
    {
        public int Current { get; private set; }
    }
    public Enumerator GetEnumerator() => throw null;
}
```

The current diagnostics are:

> error CS0117: 'C.Enumerator' does not contain a definition for 'MoveNext'
> error CS0202: foreach requires that the return type 'C.Enumerator' of 'C.GetEnumerator()' must have a suitable public 'MoveNext' method and public 'Current' property

Changing this would mean that either we just report the first, or just report the second. Both diagnostics seemed useful to me so I kept the current behavior.

Reply:
What I'm concerned about here is that the logic for reporting error is quite complicated and split across several methods. We either need to be very specific about the intentions in a comment and what the expected result is, or do a refactoring to make the logic easier to follow.